### PR TITLE
W5 - TienHuynh3 - Thread - Handler - CountDownTimer 

### PR DESCRIPTION
**Pull request by at-tienhuynh3**
#### Related issues (Optional)
Task #66

#### Description (Required)
- [x] Download image
- [x] count down timer

### CODE REVIEW

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        initData();
+        initViews();
+        initAdapters();
+    }
+
+    /**
+     * Init Data
+     */
+    private void initData() {
+        mItemList.add(new Item("aaaaa"));
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java]: WARNING: [AvoidDuplicateLiterals] The String literal "aaaaa" appears 5 times in this file; the first occurrence is on line 37
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                    while ((count = input.read(data)) != -1) {
+                        total += count;
+                        // Publish the progress
+                        mStatus = (int) (total * FINISH_PROGRESS / fileLength);
+                        output.write(data, START_PROGRESS, count);
+                        // assign bitmap
+                        mBitmap = BitmapFactory.decodeByteArray(output.toByteArray(), 0, output.toByteArray().length);
+                    }
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    private void initProgressBar() {
+        mBtnStartDownLoad.setEnabled(false);
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (mStatus < FINISH_PROGRESS) {
+                    try {
+                        Thread.sleep(500);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                            mProgressBarDownLoad.setProgress(mStatus);
+                            mTvPercentDownLoad.setText(String.valueOf(mStatus).concat("%"));
+                        }
+                    });
+                }
+                if (mStatus >= FINISH_PROGRESS) {
+                    try {
+                        Thread.sleep(1000);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/ViewImageFragment.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import android.widget.ImageView;
+
+import vn.asiantech.internship.R;
+
+/**
+ * ViewImageFragment to show image downloaded at fragment one
+ */
+public class ViewImageFragment extends Fragment {
+
+    public static ImageView mImgDownLoaded;
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/ViewImageFragment.java]: WARNING: [StaticFieldLeak] Do not place Android context classes in static fields; this is a memory leak (and also breaks Instant Run)
 A static field will leak contexts.

Non-static inner classes have an implicit reference to their outer class. If that outer class is for example a `Fragment` or `Activity`, then this reference means that the long-running handler/loader/task will hold a reference to the activity which prevents it from getting garbage collected.

Similarly, direct field references to activities and fragments from these longer running instances can cause leaks.

ViewModel classes should never point to Views or non-application Contexts.

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    private void initProgressBar() {
+        mBtnStartDownLoad.setEnabled(false);
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (mStatus < FINISH_PROGRESS) {
+                    try {
+                        Thread.sleep(500);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

**Comment by at-tienhuynh3** [comment]
@at-kietvo please review for me. Thank you!

File: app/build.gradle [hunk 7]
```diff
@@ -93,6 +93,18 @@ dependencies {
     implementation 'de.hdodenhof:circleimageview:2.2.0'
     implementation 'com.android.support:design:26.1.0'
     implementation 'me.relex:circleindicator:1.2.2@aar'
+    implementation 'com.android.support:design:26.1.0'
+    implementation 'com.android.support:support-v4:26.1.0'
+    implementation 'com.android.support:design:26.1.0'
+    implementation 'com.squareup.picasso:picasso:2.5.2'
+    implementation 'com.squareup.picasso:picasso:2.5.2'
```
Comment (at-kietvo) [inline app/build.gradle]: Duplicate.

File: app/build.gradle [hunk 8]
```diff
@@ -93,6 +93,18 @@ dependencies {
     implementation 'de.hdodenhof:circleimageview:2.2.0'
     implementation 'com.android.support:design:26.1.0'
     implementation 'me.relex:circleindicator:1.2.2@aar'
+    implementation 'com.android.support:design:26.1.0'
+    implementation 'com.android.support:support-v4:26.1.0'
+    implementation 'com.android.support:design:26.1.0'
```
Comment (at-kietvo) [inline app/build.gradle]: Duplicate.

File: app/build.gradle [hunk 9]
```diff
@@ -93,6 +93,18 @@ dependencies {
...
+    implementation 'com.squareup.picasso:picasso:2.5.2'
+    implementation 'com.squareup.picasso:picasso:2.5.2'
+    implementation 'com.android.support:design:26.1.0'
+    implementation 'de.hdodenhof:circleimageview:2.2.0'
+    implementation 'com.android.support:design:26.1.0'
+    implementation 'com.squareup.picasso:picasso:2.5.2'
+    implementation 'com.android.support:support-v4:26.1.0'
+    implementation 'com.romandanylyk:pageindicatorview:0.2.0@aar'
+    implementation 'com.romandanylyk:pageindicatorview:0.2.0'
```
Comment (at-kietvo) [inline app/build.gradle]: Duplicate.

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/adapters/ListCountDownTimerAdapter.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Override
+    public int getItemCount() {
+        return mItemLists.size();
+    }
+
+    /**
+     * This is inner class. It's used to bind data and set values to recycler view
+     */
+    class ListCountDownViewHolder extends RecyclerView.ViewHolder {
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/adapters/ListCountDownTimerAdapter.java]: Pls change to `static` class

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.threadandhandler.countdowtimer.adapters.ListCountDownTimerAdapter;
+import vn.asiantech.internship.threadandhandler.countdowtimer.models.Item;
+
+/**
+ * Created at 2017
+ * Created by jackty on 20/12/2017.
+ */
+public class CountDownTimerActivity extends AppCompatActivity {
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java]: Pls using `onDetroy()` life cycle of activity for cancel/stop/remove Count down timer after close activity.

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        public void onFinish() {
+            mTvCountDown.setText(getResources().getString(R.string.number_0));
+            mCountTime = 0;
+        }
+    };
+
+    /**
+     * This method to add 2 item after 10 second
+     */
+    CountDownTimer addItem = new CountDownTimer(10000, 1000) {
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java]: Change to `mAddItemTimer`

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                mListCountDownTimerAdapter.notifyDataSetChanged();
+                addItem.start();
+            }
+        }
+    }.start();
+
+    /**
+     * This method is used to delete middle item after 15 seconds
+     */
+    CountDownTimer deleteItem = new CountDownTimer(15000, 1000) {
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java]: Change to `mDeleteItemTimer`

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+                initProgressBar();
+                DownloadImageFromPath(LINK_IMAGE);
+            }
+        });
+    }
+
+    /**
+     * This method is used to DownLoad Image From The Internet
+     */
+    private void DownloadImageFromPath(final String path) {
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: pls change method name follow convention

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+     * this method to init percent in progress bar
+     */
+    private void initProgressBar() {
+        mBtnStartDownLoad.setEnabled(false);
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (mStatus < FINISH_PROGRESS) {
+                    try {
+                        Thread.sleep(500);
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: Why need this ?

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+        mRecyclerViewCountDownTimer.setAdapter(mListCountDownTimerAdapter);
+    }
+
+    /**
+     * Destroy Activity
+     */
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        countDownTimer.cancel();
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java]: Change the name of variable follow convention `m`

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/TabManagementActivity.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    public TabAdapter mTabAdapter;
+    private TabLayout mTabLayout;
+    private ViewPager mViewPagerContainer;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_tab_management);
+        initViews();
+        initAdapters();
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/TabManagementActivity.java]: Only Adapter pls remove `s`

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * This is count down timer
+     */
+    CountDownTimer countDownTimer = new CountDownTimer(180000, 1000) {
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java]: can you use only 1 count down timer for add , remove ?

And pls define number to `constant variable`

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.threadandhandler.countdowtimer.adapters.ListCountDownTimerAdapter;
+import vn.asiantech.internship.threadandhandler.countdowtimer.models.Item;
+
+/**
+ * Created at 2017
+ * Created by jackty on 20/12/2017.
+ */
+public class CountDownTimerActivity extends AppCompatActivity {
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java]: Pls Ctrl + A and reformat code

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+
+    private RecyclerView mRecyclerViewCountDownTimer;
+    private List<Item> mItemList = new ArrayList<>();
+    private TextView mTvCountDown;
+    private ListCountDownTimerAdapter mListCountDownTimerAdapter;
+    private int mCountTime;
+
+    @Override
+
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/countdowtimer/ui/CountDownTimerActivity.java]: Pls remove this line

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+                    connection = (HttpURLConnection) url.openConnection();
+                    connection.setDoInput(true);
+                    connection.connect();
+                    input = connection.getInputStream();
+                    output = new ByteArrayOutputStream();
+                    int fileLength = connection.getContentLength();
+                    byte data[] = new byte[1024];
+                    long total = START_PROGRESS;
+                    int count;
+                    while ((count = input.read(data)) != -1) {
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: You are limiting the part of the total for reading stream, pls remove data, we can use inputStream.read() instead.
Also remove variable of data[] and change condition of while

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+            @Override
+            public void run() {
+                HttpURLConnection connection;
+                InputStream input;
+                ByteArrayOutputStream output;
+                try {
+                    URL url = new URL(path);
+                    connection = (HttpURLConnection) url.openConnection();
+                    connection.setDoInput(true);
+                    connection.connect();
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: Pls close input, output after use
Also connection

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+                    byte data[] = new byte[1024];
+                    long total = START_PROGRESS;
+                    int count;
+                    while ((count = input.read(data)) != -1) {
+                        total += count;
+                        // Publish the progress
+                        mStatus = (int) (total * FINISH_PROGRESS / fileLength);
+                        output.write(data, START_PROGRESS, count);
+                        // assign bitmap
+                        mBitmap = BitmapFactory.decodeByteArray(output.toByteArray(), 0, output.toByteArray().length);
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: Pls move bitmap out site `while` block

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+                    }
+                } catch (IOException e) {
+                    Log.d("error", e.getMessage());
+                }
+                getActivity().runOnUiThread(new Runnable() {
+                    @Override
+                    public void run() {
+                        ((ViewImageFragment) ((TabManagementActivity) getActivity()).mTabAdapter.getItem(1)).showBitmap(mBitmap);
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: Pls check null for bitmap

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+                        @Override
+                        public void run() {
+                            mProgressBarDownLoad.setProgress(mStatus);
+                            mTvPercentDownLoad.setText(String.valueOf(mStatus).concat("%"));
+                        }
+                    });
+                }
+                if (mStatus >= FINISH_PROGRESS) {
+                    try {
+                        Thread.sleep(1000);
```
Comment (at-kietvo) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: Why sleep ?

File: app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+                    } catch (IOException e) {
+                        Log.d("error close :", e.getMessage());
+                    }
+
+                }
+                getActivity().runOnUiThread(new Runnable() {
+                    @Override
+                    public void run() {
+                        if (mBitmap != null) {
+                            if (getActivity() instanceof TabManagementActivity)
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/threadandhandler/dowloadimage/ui/DownLoadImageFragment.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

**Pull request merged**